docs: dockerize Nextflow template

yuukiiwa opened the pull request:
Fixes #{ISSUE_NUMBER}

## Checklist

- [x] I have performed a self-review of my own code
- [x] My code follows the templates/style guidelines of the repository
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my code
Here is the run of building a docker image with the four tools indicated in `environment.yml`:
```
nextflow_dsl1$ docker build . -t minimal:test
Sending build context to Docker daemon  30.21kB
Step 1/8 : FROM nfcore/base:1.13.3
 ---> 91cb6b805076
Step 2/8 : LABEL authors="your name"       description="Docker image containing all software requirements for the testing pipeline"
 ---> Running in bd3c8f41d27b
Removing intermediate container bd3c8f41d27b
 ---> 3ac215dbc48b
Step 3/8 : COPY environment.yml /
 ---> 6b223d020e16
Step 4/8 : RUN conda env create --quiet -f /environment.yml && conda clean -a
 ---> Running in 393fe0dc5fa5
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... done
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Cache location: /opt/conda/pkgs
Will remove the following tarballs:

/opt/conda/pkgs
---------------
gettext-0.19.8.1-h0b5b191_1005.tar.bz2       3.6 MB
libffi-3.3-h58526e2_2.tar.bz2                 51 KB
zstd-1.5.0-ha95c52a_0.tar.bz2                490 KB
harfbuzz-2.8.1-h83ec7ef_0.tar.bz2            1.9 MB
lz4-c-1.9.3-h9c3ff4c_0.tar.bz2               179 KB
r-base-4.0.3-h349a78a_8.tar.bz2             23.7 MB
libglib-2.68.2-h3e27bee_0.tar.bz2            3.1 MB
graphite2-1.3.13-h58526e2_1001.tar.bz2       102 KB
setuptools-49.6.0-py38h578d9bd_3.tar.bz2     929 KB
_openmp_mutex-4.5-1_gnu.tar.bz2               22 KB
bzip2-1.0.8-h7f98852_4.tar.bz2               484 KB
python-3.8.6-hffdb5ce_5_cpython.tar.bz2     26.1 MB
font-ttf-source-code-pro-2.038-h77eed37_0.tar.bz2     684 KB
python_abi-3.8-1_cp38.tar.bz2                  4 KB
gxx_impl_linux-64-9.3.0-hd87eabc_19.tar.bz2    10.7 MB
libuuid-2.32.1-h7f98852_1000.tar.bz2          28 KB
certifi-2020.12.5-py38h578d9bd_1.tar.bz2     143 KB
fontconfig-2.13.1-hba837de_1005.tar.bz2      357 KB
markdown-3.3.3-pyh9f0ad1d_0.tar.bz2           66 KB
lerc-2.2.1-h9c3ff4c_0.tar.bz2                213 KB
xorg-libsm-1.2.3-hd9c2040_1000.tar.bz2        26 KB
libnghttp2-1.43.0-h812cca2_0.tar.bz2         808 KB
gsl-2.6-he838d99_2.tar.bz2                   3.2 MB
readline-8.1-h46c0cb4_0.tar.bz2              295 KB
wheel-0.36.2-pyhd3deb0d_0.tar.bz2             31 KB
libcurl-7.76.1-hc4aaa36_1.tar.bz2            328 KB
bwidget-1.9.14-ha770c72_0.tar.bz2            119 KB
libgcc-ng-9.3.0-h2828fa1_19.tar.bz2          7.8 MB
libgfortran-ng-9.3.0-hff62375_19.tar.bz2      22 KB
libgcc-devel_linux-64-9.3.0-h7864c58_19.tar.bz2     4.0 MB
ca-certificates-2020.12.5-ha878542_0.tar.bz2     137 KB
tktable-2.10-hb7b940f_3.tar.bz2               89 KB
libwebp-base-1.2.0-h7f98852_2.tar.bz2        815 KB
libgfortran5-9.3.0-hff62375_19.tar.bz2       2.0 MB
binutils_linux-64-2.35-h67ddf6f_30.tar.bz2      22 KB
font-ttf-dejavu-sans-mono-2.37-hab24e00_0.tar.bz2     388 KB
libxcb-1.13-h7f98852_1003.tar.bz2            395 KB
libgomp-9.3.0-h2828fa1_19.tar.bz2            376 KB
cairo-1.16.0-h6cf1ce9_1008.tar.bz2           1.5 MB
jbig-2.1-h7f98852_2003.tar.bz2                43 KB
kernel-headers_linux-64-2.6.32-h77966d4_13.tar.bz2     707 KB
xorg-kbproto-1.0.7-h7f98852_1002.tar.bz2      27 KB
xorg-libxt-1.2.1-h7f98852_2.tar.bz2          375 KB
xz-5.2.5-h516909a_1.tar.bz2                  343 KB
fonts-conda-forge-1-0.tar.bz2                  4 KB
_r-mutex-1.0.1-anacondar_1.tar.bz2             3 KB
zipp-3.4.1-pyhd8ed1ab_0.tar.bz2               11 KB
liblapack-3.9.0-9_openblas.tar.bz2            11 KB
curl-7.76.1-h979ede3_1.tar.bz2               149 KB
sed-4.8-he412f7d_0.tar.bz2                   264 KB
c-ares-1.17.1-h7f98852_1.tar.bz2             109 KB
ld_impl_linux-64-2.35.1-hea4e1c9_2.tar.bz2     618 KB
libxml2-2.9.12-h72842e0_0.tar.bz2            772 KB
fribidi-1.0.10-h36c2ea0_0.tar.bz2            112 KB
gcc_impl_linux-64-9.3.0-h70c0ae5_19.tar.bz2    43.2 MB
libedit-3.1.20191231-he28a2e2_2.tar.bz2      121 KB
importlib-metadata-4.3.0-py38h578d9bd_0.tar.bz2      30 KB
fonts-conda-ecosystem-1-0.tar.bz2              4 KB
xorg-libxext-1.3.4-h7f98852_1.tar.bz2         54 KB
pymdown-extensions-8.0.1-pyh9f0ad1d_0.tar.bz2     129 KB
xorg-xproto-7.0.31-h7f98852_1007.tar.bz2      73 KB
libdeflate-1.7-h7f98852_5.tar.bz2             67 KB
pip-21.1.2-pyhd8ed1ab_0.tar.bz2              1.1 MB
gcc_linux-64-9.3.0-hf25ea35_30.tar.bz2        23 KB
libblas-3.9.0-9_openblas.tar.bz2              11 KB
xorg-renderproto-0.11.1-h7f98852_1002.tar.bz2       9 KB
gxx_linux-64-9.3.0-h3fbe746_30.tar.bz2        23 KB
zlib-1.2.11-h516909a_1010.tar.bz2            106 KB
jpeg-9d-h36c2ea0_0.tar.bz2                   264 KB
libev-4.33-h516909a_1.tar.bz2                104 KB
xorg-libxdmcp-1.1.3-h7f98852_0.tar.bz2        19 KB
pixman-0.40.0-h36c2ea0_0.tar.bz2             627 KB
gfortran_impl_linux-64-9.3.0-hc4a2995_19.tar.bz2    10.1 MB
ncurses-6.2-h58526e2_4.tar.bz2               985 KB
libtiff-4.3.0-hf544144_1.tar.bz2             668 KB
libiconv-1.16-h516909a_0.tar.bz2             1.4 MB
gfortran_linux-64-9.3.0-hdc58fab_30.tar.bz2      22 KB
sysroot_linux-64-2.12-h77966d4_13.tar.bz2    30.2 MB
libopenblas-0.3.15-pthreads_h8fe5266_1.tar.bz2     9.2 MB
sqlite-3.35.5-h74cdb3f_0.tar.bz2             1.4 MB
binutils_impl_linux-64-2.35.1-h193b22a_2.tar.bz2     9.3 MB
pcre2-10.36-h032f7d1_1.tar.bz2               1.1 MB
xorg-libxrender-0.9.10-h7f98852_1003.tar.bz2      32 KB
font-ttf-inconsolata-3.000-h77eed37_0.tar.bz2      94 KB
libcblas-3.9.0-9_openblas.tar.bz2             11 KB
pthread-stubs-0.4-h36c2ea0_1001.tar.bz2        5 KB
pcre-8.44-he1b5a44_0.tar.bz2                 261 KB
xorg-xextproto-7.3.0-h7f98852_1002.tar.bz2      28 KB
libstdcxx-devel_linux-64-9.3.0-hb016644_19.tar.bz2    14.0 MB
openssl-1.1.1k-h7f98852_0.tar.bz2            2.1 MB
xorg-libx11-1.7.1-h7f98852_0.tar.bz2         946 KB
libpng-1.6.37-h21135ba_2.tar.bz2             306 KB
font-ttf-ubuntu-0.83-hab24e00_0.tar.bz2      1.9 MB
make-4.3-hd18ef5c_1.tar.bz2                  507 KB
_libgcc_mutex-0.1-conda_forge.tar.bz2          3 KB
libstdcxx-ng-9.3.0-h6de172a_19.tar.bz2       4.0 MB
pango-1.48.5-hb8ff022_0.tar.bz2              398 KB
xorg-libxau-1.0.9-h7f98852_0.tar.bz2          13 KB
libssh2-1.9.0-ha56f1ee_6.tar.bz2             226 KB
tk-8.6.10-h21135ba_1.tar.bz2                 3.2 MB
freetype-2.10.4-h0708190_1.tar.bz2           890 KB
xorg-libice-1.0.10-h7f98852_0.tar.bz2         58 KB
icu-68.1-h58526e2_0.tar.bz2                 13.0 MB
krb5-1.17.2-h926e7f8_0.tar.bz2               1.4 MB

---------------------------------------------------
Total:                                     252.1 MB

Proceed ([y]/n)? 
Removed gettext-0.19.8.1-h0b5b191_1005.tar.bz2
Removed libffi-3.3-h58526e2_2.tar.bz2
Removed zstd-1.5.0-ha95c52a_0.tar.bz2
Removed harfbuzz-2.8.1-h83ec7ef_0.tar.bz2
Removed lz4-c-1.9.3-h9c3ff4c_0.tar.bz2
Removed r-base-4.0.3-h349a78a_8.tar.bz2
Removed libglib-2.68.2-h3e27bee_0.tar.bz2
Removed graphite2-1.3.13-h58526e2_1001.tar.bz2
Removed setuptools-49.6.0-py38h578d9bd_3.tar.bz2
Removed _openmp_mutex-4.5-1_gnu.tar.bz2
Removed bzip2-1.0.8-h7f98852_4.tar.bz2
Removed python-3.8.6-hffdb5ce_5_cpython.tar.bz2
Removed font-ttf-source-code-pro-2.038-h77eed37_0.tar.bz2
Removed python_abi-3.8-1_cp38.tar.bz2
Removed gxx_impl_linux-64-9.3.0-hd87eabc_19.tar.bz2
Removed libuuid-2.32.1-h7f98852_1000.tar.bz2
Removed certifi-2020.12.5-py38h578d9bd_1.tar.bz2
Removed fontconfig-2.13.1-hba837de_1005.tar.bz2
Removed markdown-3.3.3-pyh9f0ad1d_0.tar.bz2
Removed lerc-2.2.1-h9c3ff4c_0.tar.bz2
Removed xorg-libsm-1.2.3-hd9c2040_1000.tar.bz2
Removed libnghttp2-1.43.0-h812cca2_0.tar.bz2
Removed gsl-2.6-he838d99_2.tar.bz2
Removed readline-8.1-h46c0cb4_0.tar.bz2
Removed wheel-0.36.2-pyhd3deb0d_0.tar.bz2
Removed libcurl-7.76.1-hc4aaa36_1.tar.bz2
Removed bwidget-1.9.14-ha770c72_0.tar.bz2
Removed libgcc-ng-9.3.0-h2828fa1_19.tar.bz2
Removed libgfortran-ng-9.3.0-hff62375_19.tar.bz2
Removed libgcc-devel_linux-64-9.3.0-h7864c58_19.tar.bz2
Removed ca-certificates-2020.12.5-ha878542_0.tar.bz2
Removed tktable-2.10-hb7b940f_3.tar.bz2
Removed libwebp-base-1.2.0-h7f98852_2.tar.bz2
Removed libgfortran5-9.3.0-hff62375_19.tar.bz2
Removed binutils_linux-64-2.35-h67ddf6f_30.tar.bz2
Removed font-ttf-dejavu-sans-mono-2.37-hab24e00_0.tar.bz2
Removed libxcb-1.13-h7f98852_1003.tar.bz2
Removed libgomp-9.3.0-h2828fa1_19.tar.bz2
Removed cairo-1.16.0-h6cf1ce9_1008.tar.bz2
Removed jbig-2.1-h7f98852_2003.tar.bz2
Removed kernel-headers_linux-64-2.6.32-h77966d4_13.tar.bz2
Removed xorg-kbproto-1.0.7-h7f98852_1002.tar.bz2
Removed xorg-libxt-1.2.1-h7f98852_2.tar.bz2
Removed xz-5.2.5-h516909a_1.tar.bz2
Removed fonts-conda-forge-1-0.tar.bz2
Removed _r-mutex-1.0.1-anacondar_1.tar.bz2
Removed zipp-3.4.1-pyhd8ed1ab_0.tar.bz2
Removed liblapack-3.9.0-9_openblas.tar.bz2
Removed curl-7.76.1-h979ede3_1.tar.bz2
Removed sed-4.8-he412f7d_0.tar.bz2
Removed c-ares-1.17.1-h7f98852_1.tar.bz2
Removed ld_impl_linux-64-2.35.1-hea4e1c9_2.tar.bz2
Removed libxml2-2.9.12-h72842e0_0.tar.bz2
Removed fribidi-1.0.10-h36c2ea0_0.tar.bz2
Removed gcc_impl_linux-64-9.3.0-h70c0ae5_19.tar.bz2
Removed libedit-3.1.20191231-he28a2e2_2.tar.bz2
Removed importlib-metadata-4.3.0-py38h578d9bd_0.tar.bz2
Removed fonts-conda-ecosystem-1-0.tar.bz2
Removed xorg-libxext-1.3.4-h7f98852_1.tar.bz2
Removed pymdown-extensions-8.0.1-pyh9f0ad1d_0.tar.bz2
Removed xorg-xproto-7.0.31-h7f98852_1007.tar.bz2
Removed libdeflate-1.7-h7f98852_5.tar.bz2
Removed pip-21.1.2-pyhd8ed1ab_0.tar.bz2
Removed gcc_linux-64-9.3.0-hf25ea35_30.tar.bz2
Removed libblas-3.9.0-9_openblas.tar.bz2
Removed xorg-renderproto-0.11.1-h7f98852_1002.tar.bz2
Removed gxx_linux-64-9.3.0-h3fbe746_30.tar.bz2
Removed zlib-1.2.11-h516909a_1010.tar.bz2
Removed jpeg-9d-h36c2ea0_0.tar.bz2
Removed libev-4.33-h516909a_1.tar.bz2
Removed xorg-libxdmcp-1.1.3-h7f98852_0.tar.bz2
Removed pixman-0.40.0-h36c2ea0_0.tar.bz2
Removed gfortran_impl_linux-64-9.3.0-hc4a2995_19.tar.bz2
Removed ncurses-6.2-h58526e2_4.tar.bz2
Removed libtiff-4.3.0-hf544144_1.tar.bz2
Removed libiconv-1.16-h516909a_0.tar.bz2
Removed gfortran_linux-64-9.3.0-hdc58fab_30.tar.bz2
Removed sysroot_linux-64-2.12-h77966d4_13.tar.bz2
Removed libopenblas-0.3.15-pthreads_h8fe5266_1.tar.bz2
Removed sqlite-3.35.5-h74cdb3f_0.tar.bz2
Removed binutils_impl_linux-64-2.35.1-h193b22a_2.tar.bz2
Removed pcre2-10.36-h032f7d1_1.tar.bz2
Removed xorg-libxrender-0.9.10-h7f98852_1003.tar.bz2
Removed font-ttf-inconsolata-3.000-h77eed37_0.tar.bz2
Removed libcblas-3.9.0-9_openblas.tar.bz2
Removed pthread-stubs-0.4-h36c2ea0_1001.tar.bz2
Removed pcre-8.44-he1b5a44_0.tar.bz2
Removed xorg-xextproto-7.3.0-h7f98852_1002.tar.bz2
Removed libstdcxx-devel_linux-64-9.3.0-hb016644_19.tar.bz2
Removed openssl-1.1.1k-h7f98852_0.tar.bz2
Removed xorg-libx11-1.7.1-h7f98852_0.tar.bz2
Removed libpng-1.6.37-h21135ba_2.tar.bz2
Removed font-ttf-ubuntu-0.83-hab24e00_0.tar.bz2
Removed make-4.3-hd18ef5c_1.tar.bz2
Removed _libgcc_mutex-0.1-conda_forge.tar.bz2
Removed libstdcxx-ng-9.3.0-h6de172a_19.tar.bz2
Removed pango-1.48.5-hb8ff022_0.tar.bz2
Removed xorg-libxau-1.0.9-h7f98852_0.tar.bz2
Removed libssh2-1.9.0-ha56f1ee_6.tar.bz2
Removed tk-8.6.10-h21135ba_1.tar.bz2
Removed freetype-2.10.4-h0708190_1.tar.bz2
Removed xorg-libice-1.0.10-h7f98852_0.tar.bz2
Removed icu-68.1-h58526e2_0.tar.bz2
Removed krb5-1.17.2-h926e7f8_0.tar.bz2
WARNING: /root/.conda/pkgs does not exist
Cache location: /opt/conda/pkgs
Will remove the following packages:
/opt/conda/pkgs
---------------

python_abi-3.8-1_cp38                         10 KB
libgfortran-ng-9.3.0-hff62375_19              92 KB
libcblas-3.9.0-9_openblas                     41 KB
liblapack-3.9.0-9_openblas                    41 KB
fonts-conda-forge-1-0                         11 KB
_openmp_mutex-4.5-1_gnu                       93 KB
fonts-conda-ecosystem-1-0                      9 KB
libblas-3.9.0-9_openblas                      42 KB
pthread-stubs-0.4-h36c2ea0_1001               14 KB
_libgcc_mutex-0.1-conda_forge                  6 KB
_r-mutex-1.0.1-anacondar_1                     9 KB

---------------------------------------------------
Total:                                       368 KB

Proceed ([y]/n)? 
removing python_abi-3.8-1_cp38
removing libgfortran-ng-9.3.0-hff62375_19
removing libcblas-3.9.0-9_openblas
removing liblapack-3.9.0-9_openblas
removing fonts-conda-forge-1-0
removing _openmp_mutex-4.5-1_gnu
removing fonts-conda-ecosystem-1-0
removing libblas-3.9.0-9_openblas
removing pthread-stubs-0.4-h36c2ea0_1001
removing _libgcc_mutex-0.1-conda_forge
removing _r-mutex-1.0.1-anacondar_1
Removing intermediate container 393fe0dc5fa5
 ---> 3a3dd23a6b5d
Step 5/8 : ENV PATH /opt/conda/envs/testing/bin:$PATH
 ---> Running in 6f3e485b0a55
Removing intermediate container 6f3e485b0a55
 ---> 0bc50a4b1f35
Step 6/8 : RUN conda env export --name testing > testing.yml
 ---> Running in 066e1dbde291
Removing intermediate container 066e1dbde291
 ---> 3564c3e3700d
Step 7/8 : RUN touch .Rprofile
 ---> Running in ddda73e35b9c
Removing intermediate container ddda73e35b9c
 ---> 9ec266977873
Step 8/8 : RUN touch .Renviron
 ---> Running in 2ac3efe2f82f
Removing intermediate container 2ac3efe2f82f
 ---> 395b5836c422
Successfully built 395b5836c422
Successfully tagged minimal:test
nextflow_dsl1$ docker image ls
REPOSITORY                     TAG          IMAGE ID       CREATED          SIZE
minimal                        test         395b5836c422   11 seconds ago   1.37GB
```

mfansler commented on docs/templates/nextflow_dsl1/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+RUN conda env create --quiet -f /environment.yml && conda clean -a
+
+# Add conda installation dir to PATH (instead of doing 'conda activate')
+ENV PATH /opt/conda/envs/[execution_workflow]/bin:$PATH
```
This can potentially lead to problems. First, this particular image already has `conda activate base` in the `/root/.bashrc`, and that will prepend to `PATH` during shell sessions (i.e., take precedence over the environment). Second, not using `conda activate` fails to run activation scripts, which can lead to undefined behavior.

Better practice here would be to remove `conda activate base` and replace it with `conda activate [execution_workflow]`. I'd combine this with above to:

```bash
RUN conda env create -qf /environment.yml && \
    sed -i 's/^conda activate base$/conda activate [execution_workflow]/' ~/.bashrc && \
    conda clean -qafy
```

yuukiiwa replied:
Thanks! I have updated the line accordingly!

mfansler commented on docs/templates/nextflow_dsl1/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+
+# Instruct R processes to use these empty files instead of clashing with a local version
+RUN touch .Rprofile
+RUN touch .Renviron
```
Could be combined so Docker doesn't have to dump an intermediate.

```bash
RUN touch .Rprofile .Renviron
```

yuukiiwa replied:
I have merged them into one line and have mute them, so if people need R then they will unmute this line.

mfansler reviewed (changes requested):
Maybe the `environment.yml` should have the dependencies commented out by default? Usually, solving an environment with both `r` and `python` is difficult - unless users definitely need both, it is better to avoid it.

yuukiiwa commented:
HI @mfansler, thank you for the suggestions! I have updated them accordingly. Also, I also added a line to specify the docker container in `nextflow.config`.

mfansler reviewed (approved):
all good!
